KAFKA-14274, #4: introducing FetchBuffer and FetchCollector

kirktrue opened the pull request:
More abstraction of the `Fetcher` code to split into classes to buffer raw results (`FetchBuffer`) and to collect raw results into `ConsumerRecords` (`FetchCollector`).

kirktrue committed:
KAFKA-14274, #4: introducing FetchBuffer and FetchCollector

More abstraction of the Fetcher code to split into classes to buffer raw results (FetchBuffer) and to collect raw results into ConsumerRecords (FetchCollector).

kirktrue commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/IdempotentCloserTest.java:
```diff
@@ -28,7 +28,7 @@
 public class IdempotentCloserTest {
 
-    private static final Runnable CALLBACK_NO_OP = () -> {};
+    private static final Runnable CALLBACK_NO_OP = () -> { };
```
Reviewer's note: this was a checkstyle failure from a previous PR which I did not catch 😞

lianetm replied:
I just noticed it too and actually I think there was a 2nd one in the same file

lianetm replied:
Doubled checked, not a failure, just an extra leading space in ln 172 that Intellij noticed ;)

philipnee commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/FetchBuffer.java:
```diff
@@ -0,0 +1,145 @@
+import java.util.function.Predicate;
+
+/**
+ * {@code FetchBuffer} buffers up the results from the broker responses as they are received. It is essentially a
```
can we leave a comment about thread safety.  AFAIK this is a thread safe buffer.

kirktrue replied:
It isn't thread safe, actually 😄 

Documentation was added, but please review and un-resolve this comment if you think it is lacking.

philipnee reviewed (commented):
Thanks Kirk - QQ: Do you think we need unit test for collectFetch and FetchBuffer? Do you think the existing Fetcher tests already are testing these methods/functions.

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -98,16 +113,71 @@ public void onFailure(RuntimeException e) {
+        return fetchCollector.collectFetch(fetchBuffer);
+    }
+
+    public boolean hasCompletedFetches() {
```
Looks to me this is only used for testing so maybe leave it package-private as it was before and add the visibleForTesting comment

kirktrue replied:
`Fetcher.hasCompletedFetches()` is is used from `KafkaConsumer.poll()`. The former is in the `o.a.k.clients.consumer.internals` package and the latter is in the `o.a.k.clients.consumer` package. Since they weren't the same package I assumed that the method needed to be `public`. Weird 🤔

kirktrue replied:
I removed the `public` method qualifier, as suggested. Thanks!

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractFetch.java:
```diff
@@ -97,31 +81,13 @@ public AbstractFetch(final LogContext logContext,
         this.time = time;
     }
 
-    /**
```
Just wondering, is there a reason to move these 2 to the child class Fetcher? I see they just perform some checks on the what is now the fetcherBuffer, which is defined here in the AbstractFetcher, so I would say they still make sense here, and we would avoid an unneeded change.

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractFetch.java:
```diff
@@ -97,31 +81,13 @@ public AbstractFetch(final LogContext logContext,
-     *
-     * @return true if there are completed fetches, false otherwise
-     */
-    boolean hasCompletedFetches() {
```
Just wondering, is there a reason to move these 2 to the child class Fetcher? I see they just perform some action on what is now the FetcherBuffer, that is defined here in the AbstractFetch, so I would say it still makes sense to have the 2 functions here, and avoid an unneeded change maybe?

kirktrue replied:
Thanks for catching that. Let me try to remember why I moved them...

kirktrue replied:
I have carefully set these two methods back where I found them 😄

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/FetchBuffer.java:
```diff
@@ -0,0 +1,145 @@
+     *
+     * @return {@code true} if there are completed fetches that match the {@link Predicate}, {@code false} otherwise
+     */
+    public boolean hasCompletedFetches(Predicate<CompletedFetch<K, V>> predicate) {
```
This could be package private for now, and maybe review the comment above since it is not used for testing from what I see (if there is a test coming most probably package-private would be enough too)

kirktrue replied:
I removed the `public` qualifier from all of the methods as they're only used by internal consumer code & tests.

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/FetchBuffer.java:
```diff
@@ -0,0 +1,145 @@
+    }
+
+    /**
+     * Returns {@code true} of there are no completed fetches pending to return to the user.
```
typo "if"

kirktrue replied:
Fixed. Let me know of you find any others.

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/FetchBuffer.java:
```diff
@@ -0,0 +1,145 @@
+
+    /**
+     * Updates the buffer to retain only the fetch data that corresponds to the given partitions. Any previously
+     * {@link CompletedFetch fetched data} is its partition is not in the given set of partitions.
```
typo

kirktrue replied:
de-typo'd

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/FetchBuffer.java:
```diff
@@ -0,0 +1,145 @@
+ * @param <K> Record key type
+ * @param <V> Record value type
+ */
+public class FetchBuffer<K, V> implements Closeable {
```
We should add tests for this FetchBuffer I would say.  Part of it is covered in the FetcherTests but there's more without tests

kirktrue committed:
Updates from feedback

kirktrue committed:
Update AbstractFetch.java

kirktrue committed:
Update FetchBuffer.java

Removed public qualifier and fixed typo.

kirktrue committed:
Update FetchBuffer.java

Fixed incomprehensible attempt at documentation

kirktrue commented:
@lianetm @philipnee I think the main missing part is the unit tests for the `FetchBuffer` and `FetchCollector`. I was relying on the unit tests from `FetcherTest` to cover them, but it turns out that they're not covered fully because `AbstractFetch` wasn't `close`-ing the `FetchBuffer` 😱

kirktrue committed:
Properly calling FetchBuffer.close() on AbstractFetch.close()

philipnee reviewed (approved):
LG!

kirktrue committed:
Added unit tests and some minor refactoring for FetchBuffer

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchBufferTest.java:
```diff
@@ -0,0 +1,200 @@
+            assertTrue(fetchBuffer.hasCompletedFetches(p -> true));
+            assertFalse(fetchBuffer.isEmpty());
+            assertNotNull(fetchBuffer.peek());
+            assertEquals(completedFetch, fetchBuffer.peek());   // Reference equality
```
Mostly regarding the comment. If we do want to check that the buffer is returning the same reference we should probably use assertSame. AssertEquals could be returning true for 2 different references if we ever override the equals on the CompletedFetch, so the test wouldn't be doing what the comment hints it does. (If the intention is to check both, I would add the assertSame along with the existing assertEquals so we are covered for reference and value)

kirktrue replied:
Changed `assertEquals` to `assertSame` as suggested to more clearly convey reference comparison.

kirktrue committed:
Adding the beginning of a FetchCollector unit test

kirktrue committed:
Update FetchBufferTest.java

Changed assertEquals to assertSame to more clearly convey reference comparison.

kirktrue committed:
Update FetchCollectorTest.java

Added more tests

kirktrue committed:
Update FetchCollectorTest.java

Added another unit test

kirktrue committed:
Update FetchCollectorTest.java

More tests

kirktrue committed:
Refactoring

kirktrue committed:
Update FetchCollectorTest.java

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/FetchCollector.java:
```diff
@@ -143,7 +143,7 @@ public Fetch<K, V> collectFetch(final FetchBuffer<K, V> fetchBuffer) {
     }
 
-    private Fetch<K, V> fetchRecords(final CompletedFetch<K, V> nextInLineFetch) {
+    Fetch<K, V> fetchRecords(final CompletedFetch<K, V> nextInLineFetch) {
```
seems like this could remain private as it used to be

kirktrue replied:
I think I had designs to write unit tests against that method directly that never materialized. I'll double-check and revert back to `private` if possible. Thanks!

kirktrue replied:
Done

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/FetchCollector.java:
```diff
@@ -236,8 +235,7 @@ private CompletedFetch<K, V> initializeCompletedFetch(final CompletedFetch<K, V>
 
-    private CompletedFetch<K, V> handleInitializeCompletedFetchSuccess(final SubscriptionState subscriptions,
-                                                                       final CompletedFetch<K, V> completedFetch) {
+    CompletedFetch<K, V> handleInitializeSuccess(final CompletedFetch<K, V> completedFetch) {
```
private as it used to be?

kirktrue replied:
Done

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/FetchCollector.java:
```diff
@@ -301,9 +299,7 @@ private CompletedFetch<K, V> handleInitializeCompletedFetchSuccess(final Subscri
-    private void handleInitializeCompletedFetchErrors(final SubscriptionState subscriptions,
-                                                      final CompletedFetch<K, V> completedFetch,
-                                                      final Errors error) {
+    void handleInitializeErrors(final CompletedFetch<K, V> completedFetch, final Errors error) {
```
private as it used to be?

kirktrue replied:
Done

kirktrue committed:
Changed visibility of methods to be more private

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchBufferTest.java:
```diff
@@ -0,0 +1,201 @@
+     * Verifies {@link FetchBuffer#close()}} closes the buffered data for both the queue and the next-in-line buffer.
+     */
+    @Test
+    public void testCloseClearsData() {
```
Should we add another simple test for the close, simply to ensure that it is effectively idempotent? I know we are already testing the idempotent logic but this would ensure that we are actually applying it in the fetcher close

kirktrue replied:
OK. I'll take a look a bit later today hopefully.

lianetm reviewed (approved):
LGTM, thanks for the PR!
